iobuf: optimize equality

rockwotj opened the pull request:
I was running benchmark/fuzzer that did a lot of verification of data
and I noticed iobuf equality showing up in the benchmark profiles. This
commit optimizes that away.

Before:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.eq_bench_small         28380000    35.508ns     0.192ns    35.316ns    35.955ns       0.000       0.000       635.4       144.9
iobuf.eq_bench_medium           10000   180.590us    21.070ns   180.560us   180.706us       0.000       0.000   2580562.3    737934.3
iobuf.eq_bench_large             5000   365.385us   143.483ns   365.241us   366.229us       0.000       0.000   8300680.4   1487054.6
iobuf.cmp_bench_small        25260000    39.563ns     0.129ns    39.251ns    39.787ns       0.000       0.000       594.3       161.0
iobuf.cmp_bench_medium          10000   135.743us    17.305ns   135.711us   135.779us       0.000       0.000   2396246.6    553506.3
iobuf.cmp_bench_large            5000   436.648us    57.891ns   436.434us   504.353us       0.000       0.000   7707794.9   1836157.7
```

After:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.eq_bench_small        149070000     6.447ns     0.001ns     6.446ns     6.448ns       0.000       0.000        87.9        25.7
iobuf.eq_bench_medium          285000     3.062us     0.546ns     3.062us     3.064us       0.000       0.000     24756.9     12404.8
iobuf.eq_bench_large            90000     9.881us     1.125ns     9.877us     9.882us       0.000       0.000     79231.9     40052.9
iobuf.cmp_bench_small       149765000     6.436ns     0.001ns     6.423ns     6.438ns       0.000       0.000        83.5        25.7
iobuf.cmp_bench_medium         305000     2.844us     0.419ns     2.843us     2.845us       0.000       0.000     23203.5     11544.5
iobuf.cmp_bench_large           95000     9.596us     1.370ns     9.594us     9.603us       0.000       0.000     77184.9     38924.3
```

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
iobuf: optimize equality

I was running benchmark/fuzzer that did a lot of verification of data
and I noticed iobuf equality showing up in the benchmark profiles. This
commit optimizes that away.

Before:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.eq_bench_small         28380000    35.508ns     0.192ns    35.316ns    35.955ns       0.000       0.000       635.4       144.9
iobuf.eq_bench_medium           10000   180.590us    21.070ns   180.560us   180.706us       0.000       0.000   2580562.3    737934.3
iobuf.eq_bench_large             5000   365.385us   143.483ns   365.241us   366.229us       0.000       0.000   8300680.4   1487054.6
```

After:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.eq_bench_small        149070000     6.447ns     0.001ns     6.446ns     6.448ns       0.000       0.000        87.9        25.7
iobuf.eq_bench_medium          285000     3.062us     0.546ns     3.062us     3.064us       0.000       0.000     24756.9     12404.8
iobuf.eq_bench_large            90000     9.881us     1.125ns     9.877us     9.882us       0.000       0.000     79231.9     40052.9
```

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--06f56328-bfcf-4e91-a91d-1a2c491673ad-->
<!--TEST RESULTS-->
<!--2b7b9281-9234-4897-b0af-f0b629572db3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74148">build#74148</a></summary>


| test_class               | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                         |
|:-------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/74148#0199e269-2dcf-45d7-9dab-4b3740da83e2 | FLAKY         | 20/21    | upstream reliability is '99.53810623556582'. current run reliability is '95.23809523809523'. drift is 4.30001 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudStorageScrubberTest&test_method=test_scrubber |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74172">build#74172</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                       |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74172#0199e378-484a-4262-bce2-17088d3616a2 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write |
</details>
<!--/body-->

StephanDollberg commented on src/v/bytes/tests/iobuf_bench.cc:
```diff
@@ -31,6 +31,30 @@ size_t move_bench() {
 
+template<size_t Size>
+size_t eq_bench() {
+    iobuf a = iobuf::from(random_generators::gen_alphanum_string(Size));
```
While it's currently being done in many places using randomness here is discouraged as it will make the runtime non-deterministic so it's better to just generate two random strings (with no matching prefix) and then just hardcode them.

rockwotj replied:
Ah I thought that Travis's change fixed that. Will update

StephanDollberg replied:
It fixes like half of the problem (he will announce this better at some point in the future as he is still trying to work out some edge cases).

rockwotj replied:
I also optimized `<=>` because it was the same pattern

rockwotj committed:
iobuf/tests: deterministic benchmark

rockwotj committed:
iobuf: optimize <=>

Before:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.cmp_bench_small        25260000    39.563ns     0.129ns    39.251ns    39.787ns       0.000       0.000       594.3       161.0
iobuf.cmp_bench_medium          10000   135.743us    17.305ns   135.711us   135.779us       0.000       0.000   2396246.6    553506.3
iobuf.cmp_bench_large            5000   436.648us    57.891ns   436.434us   504.353us       0.000       0.000   7707794.9   1836157.7
```

After:

```
test                       iterations      median         mad         min         max      allocs       tasks        inst      cycles
iobuf.cmp_bench_small       149765000     6.436ns     0.001ns     6.423ns     6.438ns       0.000       0.000        83.5        25.7
iobuf.cmp_bench_medium         305000     2.844us     0.419ns     2.843us     2.845us       0.000       0.000     23203.5     11544.5
iobuf.cmp_bench_large           95000     9.596us     1.370ns     9.594us     9.603us       0.000       0.000     77184.9     38924.3
```

rockwotj commented on src/v/bytes/iobuf.cc:
```diff
@@ -127,29 +143,38 @@ bool iobuf::operator<(const iobuf& o) const {
+            if (rhs.empty()) {
+                rhs = other_next_view();
+                if (o_it == o.cend()) {
+                    break;
```
And will lead to an infinite loop.

rockwotj commented:
Force push: rebase (accident) and remove double anonymous namespaces

StephanDollberg commented on src/v/bytes/tests/iobuf_bench.cc:
```diff
@@ -10,17 +10,32 @@
+// data.
+iobuf make_iobuf(size_t size, bool reverse = false) {
+    auto gen = reverse ? std::views::iota('a', 'z')
+                       : std::views::iota('z', 'a');
```
Isn't this UB? It goes from 122 to 97?

rockwotj replied:
duh
